segmentation.py

```diff
@@ -0,0 +1,575 @@
+        if omega is not None:
+            self.__omega = omega
+
+        # Set prior for the dispersion/noise parameter.
+        if sigma is not None:
+            self.__sigma = sigma
+
+        # Set prior for the shape parameter.
+        if eta is not None:
+            self.__eta = eta
+
+        # Ensure algorithm initialises on first call to update.
+        self.__initialised = False
+
+        # If 'maxhypot' is set to none, no hypotheses will be trimmed.
+        if maxhypot > 0 or not None:
+            self.__maximum_hypotheses = maxhypot
+        else:
+            msg = "The input 'maxhypot' must be an integer greater than zero."
+            raise Exception(msg)
+
+        if minprob > 0:
+            self.__minimum_probability = minprob
+        else:
+            msg = "The input 'minprob' must be a float greater than zero."
+            raise Exception(msg)
+
+        # Allocate variables for tracking segments.
+        self.__hypotheses = list()
+        self.__index = list()
+        self.__probabilities = list()
+
+        self.__featfun = featfun if callable(featfun) else lambda x: x
+        self.__ratefun = ratefun if callable(ratefun) else lambda x: ratefun
+
+    def __initialise_algorithm(self, m, n):
+        """Initialise the Bcdm algorithm."""
+
+        # Ensure input dimensions are consistent.
+        if self.__m is None:
+            self.__m = m
+        elif self.__m != m:
+            msg = 'Expected %i dimensions in the predictor variable.' % m
+            raise Exception(msg)
+
+        # Ensure output dimensions are consistent.
+        if self.__n is None:
+            self.__n = n
+        elif self.__n != n:
+            msg = 'Expected %i dimensions in the response variable.' % n
+            raise Exception(msg)
+
+        # Set uninformative prior for the location parameter.
+        if self.__mu is None:
+            self.__mu = np.zeros([m, n])
+
+        # Set uninformative prior for the scale parameter.
+        if self.__omega is None:
+            self.__omega = np.eye(m)
+
+        # Set uninformative prior for the dispersion/noise parameter.
+        if self.__sigma is None:
+            self.__sigma = np.eye(n)
+
+        # Set uninformative prior for the shape parameter.
+        if self.__eta is None:
+            self.__eta = n
+
+        stat = MatrixVariateNormalInvGamma(self.__mu,
+                                           self.__omega,
+                                           self.__sigma,
+                                           self.__eta)
+
+        # Create the initial hypothesis, which states that the first segment is
+        # about to begin.
+        self.__update_count = 0
+        self.__hypotheses = [{'index': 0,
+                              'count': 0,
+                              'log_probability': 0.0,
+                              'distribution': stat,
+                              'log_constant': stat.logconst()}]
+
+    def __accum(self, x, y):
```

Comment:
How about `__softmax`? This is really just the binary version of the [softmax function](http://en.wikipedia.org/wiki/Softmax_function).